pkg/internal/appolly/appolly.go

```diff
@@ -16,65 +18,73 @@ import (
 	"github.com/grafana/beyla/v2/pkg/pipe/msg"
 )
 
+var errShutdownTimeout = errors.New("graceful shutdown has timed out")
+
 func log() *slog.Logger {
 	return slog.With("component", "beyla.Instrumenter")
 }
 
 // Instrumenter finds and instrument a service/process, and forwards the traces as
 // configured by the user
 type Instrumenter struct {
-	ctx     context.Context
-	config  *beyla.Config
-	ctxInfo *global.ContextInfo
+	ctx       context.Context
+	config    *beyla.Config
+	ctxInfo   *global.ContextInfo
+	tracersWg *sync.WaitGroup
+	bp        *pipe.Instrumenter
 
 	// tracesInput is used to communicate the found traces between the ProcessFinder and
 	// the ProcessTracer.
 	tracesInput *msg.Queue[[]request.Span]
 }
 
 // New Instrumenter, given a Config
-func New(ctx context.Context, ctxInfo *global.ContextInfo, config *beyla.Config) *Instrumenter {
+func New(ctx context.Context, ctxInfo *global.ContextInfo, config *beyla.Config) (*Instrumenter, error) {
 	setupFeatureContextInfo(ctx, ctxInfo, config)
+
+	tracesInput := msg.NewQueue[[]request.Span](msg.ChannelBufferLen(config.ChannelBufferLen))
+
+	bp, err := pipe.Build(ctx, config, ctxInfo, tracesInput)
+	if err != nil {
+		return nil, fmt.Errorf("can't instantiate instrumentation pipeline: %w", err)
+	}
+
 	return &Instrumenter{
 		ctx:         ctx,
 		config:      config,
 		ctxInfo:     ctxInfo,
-		tracesInput: msg.NewQueue[[]request.Span](msg.ChannelBufferLen(config.ChannelBufferLen)),
-	}
+		tracersWg:   &sync.WaitGroup{},
+		tracesInput: tracesInput,
+		bp:          bp,
+	}, nil
 }
 
 // FindAndInstrument searches in background for any new executable matching the
 // selection criteria.
 // Returns a channel that is closed when the Instrumenter completed all its tasks.
 // This is: when the context is cancelled, it has unloaded all the eBPF probes.
-func (i *Instrumenter) FindAndInstrument() (<-chan struct{}, error) {
+func (i *Instrumenter) FindAndInstrument() error {
 	finder := discover.NewProcessFinder(i.ctx, i.config, i.ctxInfo, i.tracesInput)
 	foundProcesses, deletedProcesses, err := finder.Start()
 	if err != nil {
-		return nil, fmt.Errorf("couldn't start Process Finder: %w", err)
+		return fmt.Errorf("couldn't start Process Finder: %w", err)
 	}
 
-	done := make(chan struct{})
 	// In background, listen indefinitely for each new process and run its
 	// associated ebpf.ProcessTracer once it is found.
-	wg := sync.WaitGroup{}
 	go func() {
 		log := log()
 		for {
 			select {
 			case <-i.ctx.Done():
-				log.Debug("stopped searching for new processes to instrument. Waiting for the eBPF tracers to be unloaded")
```

Comment:
I'd anyway log the messages, for debugging purposes.

Reply:
This log is also still there, just moved to Instrumenter.Stop() :)